Implement Deadline{Reader, Writer}

bassosimone opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/ndt-server-go/20)
<!-- Reviewable:end -->

bassosimone committed:
ioutil.go: sketch out DeadlineReader

bassosimone committed:
fix(ioutil.go): apply fix from reviewable

gfr10598 commented on util/ioutil.go:
```diff
@@ -0,0 +1,42 @@
+	if err != nil {
+		return count, err
+	}
+	count, err = io.ReadFull(dr.Reader, data)
```
Just pass dr, instead of dr.Reader

bassosimone committed:
fix(ioutil.go): typo

bassosimone committed:
fix(ioutil.go): fix suggested by greg

gfr10598 commented on util/ioutil.go:
```diff
@@ -0,0 +1,42 @@
+		deadline: 10.0 * time.Second,
+	}
+}
+
```
Or we could (in future) override io.Read

func (rdr *DeadlineReader) Read(p []byte) (n int, err error) {
   // maybe adjust deadline
   // ...
   // Then execute the read.
   return rdr.Reader.Read(p)
}

Then we can just directly call ReadFull(dr, data)

bassosimone committed:
Progress

bassosimone committed:
Merge remote-tracking branch 'origin/master' into feature/ioutil

bassosimone committed:
Make some progress; tests still failing

bassosimone commented:
@gfr10598 I updated this branch, even though a test is still failing. This is what I did:

1. I renamed `util/ioutil.go` to `util/network.go`; the reason of this change is that there is a `ioutil` in the standard library, so I though that maybe using the same name would be confusing; does this make sense, or would you rather me to revert the change and keep the `ioutil.go` naming?

2. driven by curiosity, I followed the earlier suggestion and tried to mock only the basic methods of the interface, so that we can basically implement better timeouts and have more flexibility

3. again, driven by curiosity, I mocked a `net.Conn` in the unit test, so to make sure that the code deals correctly when a `net.Conn` has a specific behavior

Currently, a test is failing. It seems the code does not properly route the call to a mocked connection's `Write` as I would like. I spent some time trying to unbreak this, but failed. I will perhaps try later with a fresher mind, and hopefully I'll sort it out :-).

(I start to like Go a lot.)

bassosimone commented:
Ah! After further thinking, it now occurs to me that what we actually need is:

```go
type TimedoutConn struct {
    net.Conn
    timeout time.Duration
}

// ...

func (c TimedoutConn) Read(b []byte) (int, error) {
    // set timeout
    // delegate to c.Conn.Read()
    // reset timeout
}

// same for write
```

There is actually no need to create any wrapper around a io.Reader or io.Writer when actually what we really want is the possibility of automatically having bare Read and Write calls terminate with timeout.

(This is independent of the issue that my code above has, but at this point, next time I have a couple of hours to spend on this, I'll go on to simplify the code. I'd really like to avoid adding any wrapping around the buffer, when we can obtain the same in a simpler way by wrapping a net.Conn)

bassosimone committed:
simplify the code wrapping net.Conn

bassosimone committed:
golint && go fmt

bassosimone commented:
Okay, @gfr10598, I guess this is now ready for review!

gfr10598 commented:
This definitely now looks like something we will be able to move to m-lab/go packages once we have used it a bit and are happy with it!

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[util/network.go, line 4 at r4](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20#-L4keVWyuvCVOMQqBBsV:-L4keVWyuvCVOMQqBBsW:b-iu1q12) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e28bfd4cbe4dd935b4ff42cffeb5cad0bee2559c/util/network.go#L4)):*
> ```Go
> // is free software under the Apache v2.0 License.
> 
> package util
> ```

We try to avoid naming things util if we can think of anything better / more specific.
How about netx or netext (short for net extention), or iox.  I've already started something like this in m-lab/go, with bqext, which has bigquery extensions.

There seem to be two schools of thought.  Some libraries put many files with lots of code in a single package.  We are trying to stay in the other regime - keeping code in relatively small encapsulated packages - though I'm beginning to worry about taking it too far.

Anyhow, for the short term I'd prefer to put things in packages that are fairly narrowly scoped.

---

*[util/network.go, line 13 at r4](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20#-L4kdxphX7t0yD-COtyK:-L4kdxphX7t0yD-COtyL:bxmphcj) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e28bfd4cbe4dd935b4ff42cffeb5cad0bee2559c/util/network.go#L13)):*
> ```Go
> 
> // DeadlineConn is a net.Conn where I/O operations have deadlines. They will
> // fail with an error if it takes more than a specific timeout to complete them.
> ```

s/specific timeout/specified time/

---

*[util/network.go, line 50 at r4](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20#-L4kdiN6hiwXqorDJL6Z:-L4kdiN6hiwXqorDJL6_:b-omppoq) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e28bfd4cbe4dd935b4ff42cffeb5cad0bee2559c/util/network.go#L50)):*
> ```Go
> 		return count, err
> 	}
> 	count, err = dc.Conn.Read(data)
> ```

I think you can just
  return dc.Conn.Read(data)

The deadline will be overwritten by the next call anyway.

---

*[util/network_test.go, line 16 at r4](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20#-L4kgfSwSofGDNgc01qh:-L4kgfSwSofGDNgc01qi:bdosksi) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e28bfd4cbe4dd935b4ff42cffeb5cad0bee2559c/util/network_test.go#L16)):*
> ```Go
> 
> type mockedConn struct {
> 	net.Conn
> ```

It took me a while to figure out that embedding the interface makes mockedConn compatible with the interface, but hides whether you have actually implemented it or not.

So I've found it is better to use a comment to make it clear that this is intended to implement net.Conn, but not embed net.Conn unless you are going to pass in an existing net.Conn instance.

Alternatively, if you don't pass in a net.Conn instance, this will fulfill the interface, but then if you try to call any methods that are not implemented, you will get a runtime segfault.

FYI, to ensure that you have successfully implemented the interface, you just need a test to pass an object ref of this type as a parameter to a function requiring the interface.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20#-:-L4kgLA3cU8b0InBrnU5:bpkoqm2)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  after requested changes.

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20#-:-L4kka6GD9y_K9t-oYnx:b-lct3zi)*
<!-- Sent from Reviewable.io -->

bassosimone commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[util/network.go, line 4 at r4](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20#-L4keVWyuvCVOMQqBBsV:-L4p9cVGmkoMtZK2wpvu:b-pv9igm) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e28bfd4cbe4dd935b4ff42cffeb5cad0bee2559c/util/network.go#L4)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

We try to avoid naming things util if we can think of anything better / more specific.
How about netx or netext (short for net extention), or iox.  I've already started something like this in m-lab/go, with bqext, which has bigquery extensions.

There seem to be two schools of thought.  Some libraries put many files with lots of code in a single package.  We are trying to stay in the other regime - keeping code in relatively small encapsulated packages - though I'm beginning to worry about taking it too far.

Anyhow, for the short term I'd prefer to put things in packages that are fairly narrowly scoped.
</blockquote></details>

It is not 100% clear to me whether you are suggesting:

1. to change the package of this file only

2. to also change the package of random.go

In both cases, I wonder whether the file(s) with the `netx` package should be moved in a directory named `next` (I have the feeling - correct me if I'm wrong - that it is surprising to have `util/network.go` in package, say, `netx`).

---

*[util/network_test.go, line 16 at r4](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20#-L4kgfSwSofGDNgc01qh:-L4pBDCpRuoQfHDvFt_W:bqyssbq) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e28bfd4cbe4dd935b4ff42cffeb5cad0bee2559c/util/network_test.go#L16)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

It took me a while to figure out that embedding the interface makes mockedConn compatible with the interface, but hides whether you have actually implemented it or not.

So I've found it is better to use a comment to make it clear that this is intended to implement net.Conn, but not embed net.Conn unless you are going to pass in an existing net.Conn instance.

Alternatively, if you don't pass in a net.Conn instance, this will fulfill the interface, but then if you try to call any methods that are not implemented, you will get a runtime segfault.

FYI, to ensure that you have successfully implemented the interface, you just need a test to pass an object ref of this type as a parameter to a function requiring the interface.
</blockquote></details>

Okay, I'll document it!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20)*
<!-- Sent from Reviewable.io -->

bassosimone committed:
network.go: don't bother with resetting the deadline

Based on feedback from @gfr10598.

bassosimone committed:
network_test.go: clarify and test mockedConn

If I understood the comments correctly, this should also
address more feedback from @gfr10598.

bassosimone committed:
fix(network.go): comment

bassosimone commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[util/network.go, line 13 at r4](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20#-L4kdxphX7t0yD-COtyK:-L4pKR0mSrQ9xV_cEX8O:b97hrml) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e28bfd4cbe4dd935b4ff42cffeb5cad0bee2559c/util/network.go#L13)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

s/specific timeout/specified time/
</blockquote></details>

done in aa26e97

---

*[util/network.go, line 50 at r4](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20#-L4kdiN6hiwXqorDJL6Z:-L4pKZDXbZY2qYe3cgx9:be340n0) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e28bfd4cbe4dd935b4ff42cffeb5cad0bee2559c/util/network.go#L50)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

I think you can just
  return dc.Conn.Read(data)

The deadline will be overwritten by the next call anyway.
</blockquote></details>

dne in  7f93862

---

*[util/network_test.go, line 16 at r4](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20#-L4kgfSwSofGDNgc01qh:-L4pKdUX1YgAgAe3aclw:b7k4fp9) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e28bfd4cbe4dd935b4ff42cffeb5cad0bee2559c/util/network_test.go#L16)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

Okay, I'll document it!
</blockquote></details>

See whether 250eff0 is adequate

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20)*
<!-- Sent from Reviewable.io -->

bassosimone commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[util/ioutil.go, line 26 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20#-L46m2wwxa0EkrIua5cI-r1-26:-L4pKus-k54z0qRpbPiM:bhpawtd) ([raw file](https://github.com/m-lab/ndt-server-go/blob/0cd4d2ce88bfa22ca415157e0a5925739ac28f9e/util/ioutil.go#L26)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Or we could (in future) override io.Read

func (rdr *DeadlineReader) Read(p []byte) (n int, err error) {
   // maybe adjust deadline
   // ...
   // Then execute the read.
   return rdr.Reader.Read(p)
}

Then we can just directly call ReadFull(dr, data)
</blockquote></details>

I think this is no longer relevant, since I adapted the code to net.Conn

---

*[util/ioutil.go, line 33 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20#-L46m2wwxa0EkrIua5cI-r1-33:-L4pKzXVikcrT5spFuSI:b-cmtq0t) ([raw file](https://github.com/m-lab/ndt-server-go/blob/0cd4d2ce88bfa22ca415157e0a5925739ac28f9e/util/ioutil.go#L33)):*
> I think this is no longer relevant, since I adapted the code to net.Conn
I think this is no longer relevant, since I adapted the code to net.Conn

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[util/network.go, line 4 at r4](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20#-L4keVWyuvCVOMQqBBsV:-L4sCGUcqKaMd7VqRGvS:byckap0) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e28bfd4cbe4dd935b4ff42cffeb5cad0bee2559c/util/network.go#L4)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

It is not 100% clear to me whether you are suggesting:

1. to change the package of this file only

2. to also change the package of random.go

In both cases, I wonder whether the file(s) with the `netx` package should be moved in a directory named `next` (I have the feeling - correct me if I'm wrong - that it is surprising to have `util/network.go` in package, say, `netx`).

</blockquote></details>

Generally, the package name should match the directory name, but the file name does not need to.

I would suggest for now just moving DeadlineXXX to a new package called netx, in a new folder named netx.

---

*[util/network_test.go, line 35 at r5](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20#-L4sDLD6CtysVZ7PPZZy:-L4sDLD6CtysVZ7PPZZz:b-lt1ctq) ([raw file](https://github.com/m-lab/ndt-server-go/blob/aa26e97db342ab3910041c9987562ea59ce5d85f/util/network_test.go#L35)):*
> ```Go
> // to access them will most certainly cause a panic at runtime.
> type mockedConn struct {
> 	net.Conn
> ```

I'd suggest making net.Conn a comment.  I think you have fully implemented the API.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-server-go/20)*
<!-- Sent from Reviewable.io -->

bassosimone committed:
move network stuff in netx package

bassosimone committed:
netx_test.go: wrong to put interface there

See https://github.com/m-lab/ndt-server-go/pull/20#issuecomment-364308718
